File: operation.py
import os
import numpy as np
from flask import Flask, request, jsonify, send_from_directory

# Explicitly find the absolute directory path of your project folder
PROJECT_ROOT = os.path.abspath(os.path.dirname(__file__))

# Configure Flask using absolute system paths to prevent 404 file path issues
app = Flask(__name__, static_folder=PROJECT_ROOT, template_folder=PROJECT_ROOT)

@app.route('/')
def home():
    """Serves the main landing page directly from the absolute root directory."""
    return send_from_directory(PROJECT_ROOT, 'index.html')

@app.route('/operation.html')
def operation_page():
    """Serves the calculation studio page."""
    return send_from_directory(PROJECT_ROOT, 'operation.html')

@app.route('/<path:path>')
def serve_static_files(path):
    """Serves your script.js, styles.css and other static files smoothly."""
    return send_from_directory(PROJECT_ROOT, path)

@app.route('/api/calculate-matrix', methods=['POST'])
def calculate_matrix():
    """Handles all computations sent from operation.js."""
    data = request.json
    op = data.get('operation')
    
    # Convert input list arrays into NumPy matrices
    try:
        mat_a = np.array(data.get('matrixA'), dtype=float)
        mat_b = np.array(data.get('matrixB'), dtype=float)
    except Exception:
        return jsonify({'error': 'Failed parsing matrix values to numbers.'}), 400
    
    try:
        if op == 'addition':
            res = mat_a + mat_b
            return jsonify({'result': res.tolist()})
            
        elif op == 'subtraction':
            res = mat_a - mat_b
            return jsonify({'result': res.tolist()})
            
        elif op == 'multiplication':
            res = np.dot(mat_a, mat_b)
            return jsonify({'result': res.tolist()})
            
        elif op == 'transpose':
            # Transposes both Matrix A and Matrix B into a structured dictionary
            return jsonify({
                'result': {
                    'matrixA': mat_a.T.tolist(),
                    'matrixB': mat_b.T.tolist()
                }
            })
            
        elif op == 'determinant':
            # Both matrices are strictly the same size now based on front-end config
            if mat_a.shape[0] != mat_a.shape[1]:
                return jsonify({'error': 'Determinant requires square matrices (Rows = Columns).'}), 400
            
            det_a = np.linalg.det(mat_a)
            det_b = np.linalg.det(mat_b)
            
            # Formats clean output values for both matrices back to operation.js
            return jsonify({
                'result': f"Det(A): {round(float(det_a), 4)}  |  Det(B): {round(float(det_b), 4)}"
            })
            
        else:
            return jsonify({'error': f'Unknown operation: {op}'}), 400
            
    except Exception as e:
        return jsonify({'error': str(e)}), 400

if __name__ == '__main__':
    print("🚀 MatrixOps Engine Starting Up...")
    app.run(debug=True, port=5500)

File: operation.js
document.addEventListener('DOMContentLoaded', () => {
    // Simulated state pulled from your previous dimension configuration page.
    // If not found, defaults directly to standard 3x3 shapes.
    let rowsA = parseInt(localStorage.getItem('matrixA_rows')) || 3;
    let colsA = parseInt(localStorage.getItem('matrixA_cols')) || 3;
    let rowsB = parseInt(localStorage.getItem('matrixB_rows')) || 3;
    let colsB = parseInt(localStorage.getItem('matrixB_cols')) || 3;

    // Elements Dom Links
    const containerA = document.getElementById('container-matrix-a');
    const containerB = document.getElementById('container-matrix-b');
    const tagA = document.getElementById('dim-tag-a');
    const tagB = document.getElementById('dim-tag-b');
    const resultSection = document.getElementById('result-display-section');
    const resultContent = document.getElementById('result-content');

    // UI Label Initialization
    tagA.textContent = `${rowsA} × ${colsA}`;
    tagB.textContent = `${rowsB} × ${colsB}`;

    // Generate functional Comma-separated row boxes based on properties
    function generateInputRows(container, rowCount, colCount, placeholderName) {
        container.innerHTML = '';
        for (let i = 0; i < rowCount; i++) {
            const input = document.createElement('input');
            input.type = 'text';
            input.className = 'matrix-string-row-input';
            
            // Helpful dynamically computed placeholder hint text values
            let hintValues = Array(colCount).fill('0').join(', ');
            input.placeholder = `${placeholderName} Row ${i + 1} (${hintValues})`;
            
            container.appendChild(input);
        }
    }

    // Build functional UI elements on workspace launch
    generateInputRows(containerA, rowsA, colsA, 'A');
    generateInputRows(containerB, rowsB, colsB, 'B');

    /**
     * Extracts values from text inputs, splitting on commas,
     * and converting elements to integers or floats as appropriate.
     */
    function parseMatrixStructure(container, expectedCols) {
        const inputElements = container.querySelectorAll('.matrix-string-row-input');
        let parsedMatrix = [];

        for (let i = 0; i < inputElements.length; i++) {
            let rowString = inputElements[i].value.trim();
            
            // Treat empty inputs as zeros fallback configuration
            if (!rowString) {
                parsedMatrix.push(Array(expectedCols).fill(0));
                continue;
            }

            // Split string items strictly based on user requested ',' token
            let rawTokens = rowString.split(',');
            if (rawTokens.length !== expectedCols) {
                alert(`Error: Row ${i + 1} requires exactly ${expectedCols} values separated by commas.`);
                return null;
            }

            let numericRow = rawTokens.map(token => {
                let trimmed = token.trim();
                if (trimmed === '') return 0;
                
                let num = Number(trimmed);
                if (isNaN(num)) {
                    alert(`Invalid number layout detected: "${trimmed}"`);
                    return null;
                }
                // Differentiates automatically between integer or decimal format configuration
                return Number.isInteger(num) ? parseInt(trimmed, 10) : parseFloat(trimmed);
            });

            if (numericRow.includes(null)) return null;
            parsedMatrix.push(numericRow);
        }
        return parsedMatrix;
    }

    // Setup action listener on click commands for operations buttons
    document.querySelectorAll('.op-card').forEach(card => {
        card.addEventListener('click', async () => {
            const operationType = card.getAttribute('data-operation');
            
            const matrixAData = parseMatrixStructure(containerA, colsA);
            const matrixBData = parseMatrixStructure(containerB, colsB);

            // Halt transaction logic pipeline execution if tracking validations fail
            if (!matrixAData || !matrixBData) return;

            // Prepare transaction item structure payload
            const payload = {
                operation: operationType,
                matrixA: matrixAData,
                matrixB: matrixBData,
                dimensions: {
                    rowsA: rowsA, colsA: colsA,
                    rowsB: rowsB, colsB: colsB
                }
            };

            try {
                // Submit transaction fetch call request structure to Flask route backend endpoint
                const response = await fetch('/api/calculate-matrix', {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify(payload)
                });

                const apiResult = await response.json();

                if (!response.ok || apiResult.error) {
                    alert(`Computation Error: ${apiResult.error || 'Unknown server fault.'}`);
                    return;
                }

                displayComputationResult(apiResult.result);

            } catch (err) {
                console.error(err);
                alert('Connection failure link dropped while processing computations with Flask backend server.');
            }
        });
    });

        // Display formatted results inside target viewport panel blocks
    function displayComputationResult(data) {
        resultContent.innerHTML = '';
        resultSection.classList.remove('hidden');

        // Handle dual matrix outputs (like Transpose of both A and B together)
        if (data && typeof data === 'object' && data.matrixA && data.matrixB) {
            renderSingleResultMatrix("Transpose of Matrix A:", data.matrixA);
            renderSingleResultMatrix("Transpose of Matrix B:", data.matrixB);
            return;
        }

        // Handle scalar outputs (like Determinant calculation results)
        if (!Array.isArray(data)) {
            const scalarDiv = document.createElement('div');
            scalarDiv.className = 'result-row';
            scalarDiv.style.fontWeight = '700';
            scalarDiv.textContent = `Value = ${data}`;
            resultContent.appendChild(scalarDiv);
            return;
        }

        // Handle standard single structural matrix grids rows (Addition, Subtraction, Multiplication)
        data.forEach(row => {
            const rowDiv = document.createElement('div');
            rowDiv.className = 'result-row';
            let rowText = row.map(val => Number.isInteger(val) ? val : val.toFixed(2)).join('   ');
            rowDiv.textContent = `[  ${rowText}  ]`;
            resultContent.appendChild(rowDiv);
        });
    }

    // Helper utility function to render clean titled matrices inside the output area
    function renderSingleResultMatrix(titleText, matrixArray) {
        const titleEl = document.createElement('div');
        titleEl.style.fontSize = '14px';
        titleEl.style.color = '#715530';
        titleEl.style.fontWeight = '600';
        titleEl.style.marginTop = '12px';
        titleEl.style.marginBottom = '6px';
        titleEl.style.textAlign = 'center';
        titleEl.textContent = titleText;
        resultContent.appendChild(titleEl);

        matrixArray.forEach(row => {
            const rowDiv = document.createElement('div');
            rowDiv.className = 'result-row';
            let rowText = row.map(val => Number.isInteger(val) ? val : val.toFixed(2)).join('   ');
            rowDiv.textContent = `[  ${rowText}  ]`;
            resultContent.appendChild(rowDiv);
        });
    }

    // Handle systemic data field clearing inputs 
    document.getElementById('btn-clear-values').addEventListener('click', () => {
        document.querySelectorAll('.matrix-string-row-input').forEach(input => input.value = '');
        resultSection.classList.add('hidden');
        resultContent.innerHTML = '';
    });
});
